File: 1D_DP/LC_213.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.12.05

class Solution {
public:

    int rob1(vector<int>& nums) {
        int n = nums.size();
        if (n == 1){
            return nums[0];
        }
        vector<int> dp(n+1, 0);
        dp[1] = nums[0];
        for (int i = 2; i <= n; i++){
            dp[i] = max(dp[i-1], dp[i-2] + nums[i-1]);
        }
        return dp[n];
    }

    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 1){
            return nums[0];
        }
        vector<int> part1, part2;
        for(int i = 0; i < n; i++){
            if (i == 0){
                part1.push_back(nums[i]);
            }
            else if (i == n-1){
                part2.push_back(nums[i]);
            }
            else{
                part1.push_back(nums[i]);
                part2.push_back(nums[i]);
            }
        }
        return max(rob1(part1), rob1(part2));
    }
};

File: 1D_DP/LC_213.py
# Written by: Jian-Yuan Lin
# Date: 2022.12.05

class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        def rob_1(nums):
            
            rob1, rob2 = 0, 0
            n = len(nums)
            #[rob1, rob2, n, n+1....]
            for i in range(n):
                temp = max(rob1 + nums[i], rob2)
                rob1 = rob2
                rob2 = temp
            return rob2
        
        return max(rob_1(nums[:-1]), rob_1(nums[1:]))